### PR TITLE
[SPARK-20213][SQL] Fix DataFrameWriter operations in SQL UI tab

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Currently the `DataFrameWriter` operations have several problems:

1. non-file-format data source writing action doesn't show up in the SQL tab in Spark UI
2. file-format data source writing action shows a scan node in the SQL tab, without saying anything about writing. (streaming also have this issue, but not fixed in this PR)
3. Spark SQL CLI actions don't show up in the SQL tab.

This PR fixes all of them, by refactoring the `ExecuteCommandExec` to make it have children.

 close https://github.com/apache/spark/pull/17540

## How was this patch tested?

existing tests.

Also test the UI manually. For a simple command: `Seq(1 -> "a").toDF("i", "j").write.parquet("/tmp/qwe")`

before this PR:
qq20170523-035840 2x
after this PR:
qq20170523-035708 2x

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rdblue , I think this issue is important and we should probably fix it in 2.2, so I take over without the streaming changes to make this PR safer to merge.

Also cc @zsxwing

**Comment by rdblue** [comment]
Agreed, sorry I haven't updated it. I was out most of last week. I'll get this fixed up as soon as I can. Thanks for all your help!

**Comment by viirya** [comment]
This seems also related to my PR #17998.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala [hunk 1]
```diff
@@ -96,11 +96,12 @@ case class AnalyzeColumnCommand(
       attributesToAnalyze.map(ColumnStat.statExprs(_, ndvMaxErr))
 
     val namedExpressions = expressions.map(e => Alias(e, e.toString)())
-    val statsRow = Dataset.ofRows(sparkSession, Aggregate(Nil, namedExpressions, relation)).head()
+    val statsRow = Dataset.ofRows(sparkSession, Aggregate(Nil, namedExpressions, relation))
+      .queryExecution.executedPlan.executeTake(1).head
```
Comment (wzhfy) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]:
```
new QueryExecution(sparkSession, Aggregate(Nil, namedExpressions, relation)).executedPlan.executeTake(1).head
```

**Comment by cloud-fan** [comment]
@viirya I'm not very clear about the purpose of the "write operator", does my PR cover it?

**Comment by viirya** [comment]
@cloud-fan The purpose of the operator is just used to record metrics of writing data. I said it's related because this PR is also going to show the writing op in UI. But looks like this change doesn't do anything about the metrics.

There are many classes are both covered by this PR and #17998, e.g. the commands and how to trigger them. I'd rebase my PR once this is merged.

**Comment by viirya** [comment]
@cloud-fan Actually, if we decide to show individual writing commands in UI like this, we may not need "write operator". To have a writing operator on top on those commands might looks confusing. Then we can directly attach the metrics to the commands.

@rxin What do you think?

**Comment by rxin** [comment]
That works too, if we can attach metrics to these commands.

**Comment by viirya** [comment]
@rxin OK. I will rebase my PR after this gets merged.

**Comment by cloud-fan** [comment]
cc @zsxwing for review

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala [hunk 2]
```diff
@@ -86,12 +86,10 @@ private[kafka010] object KafkaWriter extends Logging {
       topic: Option[String] = None): Unit = {
     val schema = queryExecution.analyzed.output
     validateQuery(queryExecution, kafkaParameters, topic)
-    SQLExecution.withNewExecutionId(sparkSession, queryExecution) {
```
Comment (viirya) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala]: After removing this, can `KafkaSink` still track its streaming writing?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 3]
```diff
@@ -426,36 +418,35 @@ case class DataSource(
...
+      fileFormat = format,
+      options = options,
+      query = data,
+      mode = mode,
+      catalogTable = catalogTable,
+      fileIndex = fileIndex)
   }
 
   /**
    * Writes the given [[DataFrame]] out to this [[DataSource]] and returns a [[BaseRelation]] for
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Comment needs to change: given [[LogicalPlan]].

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 4]
```diff
@@ -390,7 +391,8 @@ case class DataSource(
   /**
    * Writes the given [[DataFrame]] out in this [[FileFormat]].
    */
-  private def writeInFileFormat(format: FileFormat, mode: SaveMode, data: DataFrame): Unit = {
+  private def planForWritingFileFormat(
+      format: FileFormat, mode: SaveMode, data: LogicalPlan): LogicalPlan = {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: given [[LogicalPlan]].

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 5]
```diff
@@ -464,18 +455,18 @@ case class DataSource(
   }
 
   /**
-   * Writes the given [[DataFrame]] out to this [[DataSource]].
+   * Returns a logical plan to write the given [[DataFrame]] out to this [[DataSource]].
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: the give [[LogicalPlan]].

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala (hunk 2 above)
Comment (cloud-fan) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala]: yea, because it's used in `KafkaProvider` which is run with `SaveIntoDataSourceCommand`, and all commands will be wrapped by `SQLExecution.withNewExecutionId`

**Review by zsxwing — COMMENTED** [review]
Looks good to me. Just two minor questions.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 6]
```diff
@@ -965,14 +965,20 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
   }
 
   test("sanity test for SPARK-6618") {
-    (1 to 100).par.map { i =>
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: this test was failing and complained that the execution id is already set. I don't know the detail about how we schedule the 100 tasks on threads with `par`, so I changed it to use `Thread` directly to be more clear. I'll reduce the number of threads.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala (hunk 2 above)
Comment (viirya) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala]: If you mean `KafkaSourceProvider`, is it the same code path as `KafkaSink`? In `KafkaSink.addBath`, `KafkaWriter.write` is also called to write into Kafka.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 7]
```diff
@@ -426,36 +418,35 @@ case class DataSource(
...
-        query = data.logicalPlan,
-        mode = mode,
-        catalogTable = catalogTable,
-        fileIndex = fileIndex)
-      sparkSession.sessionState.executePlan(plan).toRdd
+    InsertIntoHadoopFsRelationCommand(
+      outputPath = outputPath,
+      staticPartitions = Map.empty,
+      ifPartitionNotExists = false,
+      partitionColumns = partitionColumns.map(UnresolvedAttribute.quoted),
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Is it safe to use unresolved attributes here?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: `InsertIntoHadoopFsRelationCommand` will pass `partitionColumns` to `FileFormatWriter`. Seems `FileFormatWriter` won't resolve those `partitionColumns`?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Analyzer will resolve them in `InsertIntoHadoopFsRelationCommand`.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala (hunk 2 above)
Comment (cloud-fan) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala]: `KafkaSink.addBath` is also wrapped with `SQLExecution.withNewExecutionId` in `StreamExecution`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 8]
```diff
@@ -294,32 +294,31 @@ private[hive] class TestHiveSparkSession(
         "CREATE TABLE src1 (key INT, value STRING)".cmd,
         s"LOAD DATA LOCAL INPATH '${quoteHiveFile("data/files/kv3.txt")}' INTO TABLE src1".cmd),
       TestTable("srcpart", () => {
-        sql(
-          "CREATE TABLE srcpart (key INT, value STRING) PARTITIONED BY (ds STRING, hr STRING)")
+        "CREATE TABLE srcpart (key INT, value STRING) PARTITIONED BY (ds STRING, hr STRING)"
+          .cmd.apply()
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: use `xxx.cmd.apply` instead of `sql("xx")` here, to avoid setting the execution id.

**Comment by rdblue** [comment]
@cloud-fan, can you summarize how this differs from the original PR #17540?

I have time to pick this up again, but I thought that the other PR only needed two changes:

* Merge your changes to unify the physical plans when using `ExecuteCommandExec`
* Move `withNewExecutionId` to stremaing batches, rather than the entire session

This seems to have gone though a lot of updates for testing, though, and I don't see my original commits... is this a reimplementation of those changes?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 6 above)
Comment (zsxwing) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Oh, my hunch is we probably leak execution id  to some thread in a thread pool. It's better to fix it rather than changing the test to hide it.

**Comment by cloud-fan** [comment]
@rdblue the difference:
1. unify the physical plans when using `ExecuteCommandExec`
2. Move `withNewExecutionId` to stremaing batches, rather than the entire session
3. instead of `SQLExecution.nested`, introduce some internal actions that won't set the execution id in nested exeution

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 6 above)
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: The weird thing is I can't reproduce it locally, and I'd say it's not introduced by this PR, as previously no execution id is set for commands.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala (hunk 7 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Ok. I see.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 9]
```diff
@@ -242,9 +242,12 @@ class Dataset[T] private[sql](
    * @param vertical If set to true, prints output rows vertically (one line per column value).
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Seems to add a `@param` for `isInternal` is better?

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Comment by yhuai** [comment]
I just case across this pr. I have one general feedback. It will be great if we can make a pr have a single purpose. This pr contains different kinds of changes in order to fix the UI. If refactoring is needed, I'd recommend to have separate PR for refactoring purposes. Then, use a different PR to do the actual fix.

**Comment by cloud-fan** [comment]
@yhuai thanks for the remind! I'll send separated PR for refactor in the future :)

**Comment by rdblue** [comment]
Refactor?

I thought that was the problem with the [original PR](https://github.com/apache/spark/pull/17540); that PR was too narrow and didn't unify the physical plans to get all metrics working.

**Comment by yhuai** [comment]
My suggestion was about getting changes on the interfaces of ExecutedCommandExec and SaveIntoDataSourceCommand to separate prs. It will help code review (both speed and quality).
